scheduler/system_util.go

```diff
@@ -291,6 +287,66 @@ func diffSystemAllocsForNode(
 	return result
 }
 
+// ensureSingleSystemAlloc enforces the invariant that a system or sysbatch job
+// should never have more than one allocation in a desired-running state.
+func ensureSingleSystemAlloc(result *diffResult) {
```

Comment:
Note for reviewers: I don't _love_ this implementation but because we can't rely on the order of allocations being passed in to `diffSystemAllocsForNode` (because they come from a map), it's hard to make decisions between loop iterations without hugely expanding a loop that's already huge and hard to follow. This possibly costs us a couple of extra slice memory allocations, but I think its worth it to keep it understandable?